src/binary-encoding.adoc

```diff
@@ -2298,16 +2298,302 @@ considered "shapes" rather than types because while their encoding is always sta
+then E-expressions invoking that macro will begin with an _argument encoding bitmap_ (AEB).
+An AEB is a series of bits that correspond to a macro parameter and communicate additional
+information about how the arguments corresponding to that parameter have been encoded in the current E-expression.
+In particular, the AEB indicates whether a parameter that accepts `(:void)` has any arguments at all, and how a
+grouped parameter's arguments have been delimited.
+
+The number of bits allotted to each parameter is determined by its cardinality, as shown in the table below; each
+parameter can have 0, 1, or 2 bits.
+
+[cols="^.^2a,^.^2a,^.^2a,^.^1a,^.^1a,.^6a"]
+|===
+|Grouping Mode|Cardinality|Example parameter signature|Number of bits|Bit(s) value|Encoding
+
+.6+|Ungrouped
+|Exactly-one
+|`(x int!)`
+|0
+|_n/a_
+|One expression
+
+.2+|Zero-or-one
+.2+|`(x int?)`
+.4+|1
+|`0`
+<|No expression; equivalent to `(:void)`
+
+|`1`
+<|One expression
+
+.2+|Zero-or-more
+.2+|`(x int*)`
+|`0`
+<|No expression; equivalent to `(:void)`
+
+|`1`
+<|One expression
+
+|One-or-more
+|`(x int+)`
+|0
+|_n/a_
+<|One expression
+
+.8+|Grouped
+.4+|Zero-or-more
+.4+|`(x [int])` +
+`(x int...)`
+.8+|2
+|`00`
+|No expression; equivalent to `(:void)`
+
+|`01`
+<|One expression
+
+|`10`
+<|Length-prefixed expression group
+
+|`11`
+<|Delimited expression group
+
+.4+|One-or-more
+.4+|`(x [int]+)` +
+`(x int...+)`
+|`00`
+<|_Illegal._ One-or-more forbids `(:void)`.
+
+|`01`
+<|One expression
+
+|`10`
+<|<<length_prefixed_expression_groups, Length-prefixed expression group>>
+
+|`11`
+<|<<delimited_expression_groups, Delimited expression group>>
+
+|===
+
+The total number of bits in the AEB can be calculated by analyzing the signature of the macro being invoked.
+If the macro has no parameters or all of its parameters have a cardinality of either exactly-one or one-or-more,
+no bits are required; the AEB will be omitted altogether.
+If the macro has many parameters with a cardinality other than exactly-one, it is possible for the AEB to require more
+than one byte to encode; in such cases, the bytes are written in little-endian order.
+AEB bytes can contain unused bits.
+
+Bits are assigned to the parameters in a macro's signature from left to right.
+Bits are assigned from least significant to most significant (commonly: right-to-left).
+
+[cols="<.^8a,^.^3a,^.^2a"]
+|===
+|Example parameter sequence |Bit assignments |Total bits
+
+| `()`
+|_No AEB_
+|0
+
+| `\((a int!) (b string!) (c float!))`
+|_No AEB_
+|0
+
+| `\((a int!) (b string!) (c float?))`
+| `-------c`
+| 1
+
+| `\((a int!) (b string?) (c float!))`
+| `-------b`
+| 1
+
+| `\((a int!) (b string*) (c float?))`
+| `------cb`
+| 2
+
+| `\((a int*) (b string!) (c [float]))`
+| `-----cca`
+| 3
+
+| `\((a int*) (b [string]) (c [float]))`
+| `---ccbba`
+| 5
+
+| `\((a [int]) (b [string]) (c [float]+))`
+| `--ccbbaa`
+| 6
+
+| `\((a int*) (b [string]) (c [float]) (d [bool]) (e blob...))`
+| `eddccbba` +
```

Comment:
At one point we had discussed intermingling these bits with the argument expressions, to avoid an expensive full-expression prefix.   I am still worried about the cost of having to compute all of this before rendering anything at all.

Reply:
Expanding on this a bit:
* The width of the AEB is statically known, so bytes can be reserved in the encoding buffer without needing to adjust the size later.
* 8 bytes of AEB can be stored in an integer on 64-bit architectures, which means that doing all of the bit-twiddling needed to configure the AEB is fast and easy unless you have more than 32 grouped parameters (or even more less expensive parameters).
* Keeping the AEB bytes together means you don't have to worry about splitting two bits across bytes that are potentially very far apart in memory.